src/Psl/Fun/Internal/LazyEvaluator.php

```diff
@@ -0,0 +1,42 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Fun\Internal;
+
+/**
+ * @internal
+ * @template T
+ */
+final class LazyEvaluator
+{
+    /**
+     * @var T|null
+     */
+    private mixed $instance = null;
+
+    /**
+     * @var callable(): T
+     */
+    private $initializer;
+
+    /**
+     * @param callable(): T $initializer
+     */
+    public function __construct(callable $initializer)
+    {
+        $this->initializer = $initializer;
+    }
+
+    /**
+     * @return T
+     */
+    public function __invoke(): mixed
+    {
+        if (!$this->instance) {
```

Comment:
what if the initializer is returning null?

i think it's better to have another property `bool $initialized`.

e.g:
```
$get_user = Fun\lazy(function(): ?User {
  $user_id = $this->session->get('id');
  return $this->users->findOne($id);
});
```

Reply:
Covered that case now.
Why not `throw UserNotFoundException` or return `NullUser` ? ;)

Reply:
just an example :p don't overthink it hehehe

Reply:
I was just joking :)